No more message spam when you witness someone being attacked or shot

phil235 opened the pull request:
# **YOU STILL SEE ATTACKS/SHOTS  DAMAGING YOU**
- Visible_message() when being attacked by melee or projectile, or when firing a gun, is now only viewed by the targeted mob <b>and the witnesses who are less than 3 tiles away</b>, the attacker doesn't see his own attack messages. Some visible_message() still remain unchanged for certain special effects, mostly because I don't want to have to make a visual effect for those.
- To keep things balanced melee attacks (punches, alien attacks, animal attacks, etc...) now show a visual effect (similar to what we currently have with item attacks.
- Fixes muzzle flash effect not appearing for mech guns.
- Fixes muzzle flash effect not appearing for certain ranged animal.
- Fixes the item attack effect not being visible to camera viewers and mech occupants.
- Fixes toy guns foam dart not dropping on the ground when shooting a mob on the same tile as us.
- Fixes toy smg magazine sprite being invisible sometimes.
- Fixes foam dart not facing the correct direction when fired.
- Changes the bullet projectile sprite to have a slight tracer effect so as to be more visible.
- Fixes lasereyes' laser projectile being sideways.
- fixes some other small bugs.
- I added two arguments to visible_message(), one to specify the visibility range, and one to ignore a specific mob that shouldn't get the message.

:cl: phil235
add: Visual effects appear on the target of an attack, similar to the item attack effect.
del: You now only see a message in chat when witnessing someone else getting attacked or hit by a bullet if you are close enough to the action.
/:cl:

phil235 committed:
Visible_message() when being attacked by melee or projectile is replaced by a simple message only to the target instead of all viewers.
The message when firing a gun is removed.
To keep things balanced melee attacks (punches, alien attacks, animal attacks, etc...) now show a visual effect (similar to what we current have with item attacks.
Fixes muzzle flash effect not appearing for mech guns.
Fixes muzzle flash effect not appearing for certain ranged animal.
Fixes the item attack effect not being visible to camera viewers and mech occupants.
Fixes toy guns foam dart not dropping on the ground when shooting a mob on the same tile as us.
Fixes toy smg magazine sprite being invisible sometimes.
Fixes foam dart not facing the correct direction when fired.
Changes the bullet projectile sprite to have a slight tracer effect so as to be more visible.

phil235 committed:
some tiny fixes

PKPenguin321 commented:
I like the messages though, they're good to have real feedback

Militaires commented:
i really really like the messages, it's not spammy at all dont remove that good shit dud fam pls

TechnoAlchemisto commented:
Yeah, sorry to say but having multiple entries for multiple shots is good, as well as the gun firing messages.

That's part of what makes silencers so desirable, since they don't give off the firing messages.

MrStonedOne commented:
make it a config!

GunHog commented:
Combat log tab for the chat window, maybe?

phil235 committed:
Fixing laser eye's laser being sideways.

phil235 committed:
Merge branch 'master' of https://github.com/tgstation/-tg-station into hit_message

# Conflicts:
#	code/game/mecha/mecha_defense.dm

Dorsisdwarf commented:
Can you edit things previously printed to the log?
If so, do what DF does and have multiple identical announcements just increment (IE 
"Urist McMurderboner fires the LMG" (x4) ->"Urist McMurderboner fires the LMG" (x5)

CyberbossJHCB commented:
Why don't we make this a preference, there's really good info in the spam

ShadowDeath6 commented:
Were people really being annoyed by the 5 or 6 messages that were created when somebody shot a gun or esworded someone to death?

WJohn commented:
I think this is a good thing. It's better if we move it away from text logs because those can't really be erased from a player's "memory". Many times as someone who has been attacked by someone I find myself scrolling back up to find information I would have otherwise not noticed or forgotten. Feedback happening on screen still helps you figure out what's going on in a fight, but doesn't give you pretty unfair memory in what is effectively a lynching game.

Cheridan commented:
This really isn't something that can be made a preference because it's something that has an impact on balance. Nobody is going to want to turn it off because if they do they'll miss "TRAITOR HAS HIT X WITH THE ENERGY SWORD" and suffer a penalty for it.

Anyway, yeah, what WJ said. We pretend to be a game about subterfuge and misdirection and then spit out infallible combat logs so that everyone knows exactly who pushed them that one time 10 minutes ago.

TechnoAlchemisto commented:
Wouldn't the solution to that be remove the name of the person who fired the gun or who hit you? 

I know other places do things where they say 'You hear a gunshot'

You could just scrub the name instead of having it only register once.

PKPenguin321 commented:
Or have combat logs only given to you if you're within a certain distance of the combat, like whispering

Or just don't merge this at all since the game works fine without it

Cheridan commented:
It's just ridiculous though @TechnoAlchemisto. **No other game** has messages like that, with the exception of roguelikes like Nethack where it is a requirement because is no audio to begin with.

TechnoAlchemisto commented:
Why do we have to be like other games?

People like the current system and if you want to improve it to simulate not being able to know exactly who did things, you would just have to get rid of the names.

CyberbossJHCB commented:
>SS13 
>Similar to other games

phil235 commented:
>Wouldn't the solution to that be remove the name of the person who fired the gun or who hit you? 

I don't see the point of this. If you are the victim, you will still see the attack message. If you are a bystander and the message you receive only shows the victim's name how is that better than actually looking at the visual effects that appear on the victim? Is it really that important to know the name of the guy who got shot that was running in the hallway and you saw for less than a second?

TechnoAlchemisto commented:
Attack messages are more noticeable than overlays because anyone who's played this game for an extended period of time is conditioned to notice changes in the text log.  It's also important to know what you're getting hit by, because being shot by a pellet and being shot by a bullet have the same visual effect. 

I see what you're both saying, about text outputs in a visual game being kind of pointless, but it's how ss13 is and I think most if not all players like what the combination brings to the table.

Tacolizard commented:
1. like what @TechnoAlchemisto said this would essentially be a silencer nerf
2. What SHOULD be removed is getting this spammed at you as a ghost

phil235 commented:
> It's also important to know what you're getting hit by

The title of the PR: "No more message spam when you <b>witness</b> someone being attacked or shot"

TechnoAlchemisto commented:
Oh, yeah I misread that my bad. I do think it's important to know how many times someone gets hit by something from a role-playing perspective. 

But that is way less important than what I thought this was.

ShadowDeath6 commented:
Also I actually think that chat confirmation is a better tool for new players, since half the time they could hear a bullet fire and not realize that they're being shot at or something, provided the shooter isn't very good at landing the first shot.

nullbear commented:
So then essentially this makes every weapon a parapen for the purpose of stealth attacks.

CyberbossJHCB commented:
Not to mention the syringe gun buffs

nullbear commented:
This also kills player logs.

optimumtact commented:
None of you actually even read the PR did you?

You still see attacks targeted at **you**, and all logs are still saved for admins to retrieve.

This is about messages from other people being shot/hit/bludgeoned.

ShadowDeath6 commented:
>shoot past the person

GunHog commented:
In the small chance that you included me in that question, I did read the PR and I understand that this removes messages from bystanders only. Honestly, it sounds like a nightmare for a Security player trying to find out who shot whom. Also, this means that someone who is shooting AT you but failing to hit is also more difficult to track.

PKPenguin321 commented:
>So basically it's better this way because we learned to get used to the shitty system and fuck the newbies and game design?

Nobody's forcing you to read the text in the box, phil.

I think a good compromise might be making the background of the text box flash red when you take damage or something so that indicator is still there. Is this possible

optimumtact commented:
@PKPenguin321 LEARN TO READ

GunHog commented:
@PKPenguin321 Shots that hit you are still provided in your chat box, as well as melee attacks. What is being removed is the message of someone firing a weapon and hitting someone else. A significant amount of information is being lost regardless though (which is the unfortunate intent). For example, finding a guy that shot you from a crowd, provided you were not paying full attention, will be much more difficult.

nullbear commented:
I still want to witness stuff that's going on. If i see someone get attacked by someone, i want to know to avoid the attacker. 

Being unable to witness people being attacked has quite a phenominal balance implication, that being that you can attack people and usually get away with it so long as you're not being watched directly. While this imitates realism, it has a serious effect on gameplay and balance, especially for identifying antags/killers.

kevinz000 commented:
Make it a pref because it helps me massively to have chat logs to check when there's a huge brawl. You can't scroll your video up.

pudl commented:
this shit sucks you fag

optimumtact commented:
We're not going to make everything into a pref just to satisfy every single person.

That kind of behaviour went on long enough when I wasn't a maintainer.

It leads to horribly hard to maintain code.

ShadowDeath6 commented:
>highly controversial feature added, multiple players prefer either side
no prefs tho lel

imo this is probably the best thing to have a pref for since it's entirely up to the player whether or not they want to have an important feature of the game or not

Fox-McCloud commented:
Strongly disagree with this....as numerous others have pointed out, not being able to see who attacked who is going to generate a lot of confusion in a lot of circumstances/scenarios.

Granted, while there's "visual feedback", this is less than perfect, and sometimes you need to see just who did what in an environment or situation.


Given this is visible event, it makes sense that everyone can see/hear it that's within range---this is a step in the *wrong* direction, IMO.

Razharas commented:
There is visual indication who is attacking who and even with what, and also sound indication to distinguish toys from real stuff and so on
I think its absolutely enough for immidiate gameplay, if you think you need stuff to check out for investigation the logs can be dumped into some specific "memory" log not unlike notes(which are so unused i think notes can be this memory just alright)
I fully support this idea and if there are issues with stuff not being obvious from now on we will need to solve it by adding visual and sound feedback, not chat messages

Razharas reviewed (approved):
Havent read a single note of the code but the idea itself is right

phil235 commented on code/game/atoms_movable.dm:
```diff
@@ -348,3 +348,64 @@
+	if(!I)
+		return
+
+	flick_overlay(I, clients, 5) // 5 ticks/half a second
```
I'm really not 100% about how efficient using clients list instead of finding mobs with viewers is. I did some profile with just 1 and 2 players and although the new method is faster for these cases it seems it might be way more expensive when 100 players are connected. So we might need to profile it when it's testmerged.

CyberbossJHCB commented:
@Razharas That's what the up thumb is for, not reviews

phil235 committed:
Made some slightly less uglier sprites for attack effects.dmi
Melee attacks, gun firing, and things being hit by bullets will now still give witnesses a message, but only if they are less than 3 tiles away. The attacker still doesn't get the message for its own attacks.

I added two arguments to visible_message(), one to specify the visibility range, and one to ignore a specific mob that shouldn't get the message.

phil235 commented:
I made a compromise: witnesses will still see the hit messages if they are close enough to the action (less than 3 tiles away).

phil235 committed:
derp

nullbear commented:
I have changed my vote.

CyberbossJHCB commented:
Agreed, this is MUCH better

> On Oct 15, 2016, at 4:38 PM, nullbear <notifications@github.com> wrote:
> 
> I have changed my vote.
> 
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

GunHog commented:
This is acceptable now.

KorPhaeron commented:
This is going to encourage a lot of confusion and grey areas in the rules, going to be interesting to see how admins handle players having very plausible excuses for having no idea whats going on escalation wise

MrStonedOne commented:
Update the changelog

PKPenguin321 commented:
Still on the fence about this. Needs extensive test merging if we're going to consider it.

phil235 commented:
changelog updated.
